### PR TITLE
fix: chainlink for target

### PR DESCRIPTION
**Pull request by sogipec**

### CODE REVIEW

File: contracts/transmuter/libraries/LibOracle.sol [hunk 1]
```diff
@@ -47,13 +57,18 @@ library LibOracle {
...
+        ) = _parseOracle(oracleConfig);
         if (readType == OracleReadType.EXTERNAL) {
-            ITransmuterOracle externalOracle = abi.decode(data, (ITransmuterOracle));
+            ITransmuterOracle externalOracle = abi.decode(readData, (ITransmuterOracle));
             return externalOracle.readBurn();
         }
-        uint256 _targetPrice = targetPrice(targetType);
-        oracleValue = read(readType, _targetPrice, data);
+        uint256 _targetPrice = targetPrice(targetType, targetData);
+        oracleValue = read(readType, _targetPrice, readData);
```
Comment (GuillaumeNervoXS) [inline contracts/transmuter/libraries/LibOracle.sol:71]: do we want necessarily to have the same structure as chainlink read?

File: contracts/transmuter/libraries/LibOracle.sol [hunk 2]
```diff
@@ -21,24 +21,34 @@ library LibOracle {
     /// @notice Reads the oracle value used during a redemption to compute collateral ratio for `oracleConfig`
     /// @dev This value is only sensitive to compute the collateral ratio and deduce a penalty factor
     function readRedemption(bytes memory oracleConfig) internal view returns (uint256) {
-        (OracleReadType readType, OracleTargetType targetType, bytes memory data) = _parseOracle(oracleConfig);
+        (
+            OracleReadType readType,
+            OracleReadType targetType,
+            bytes memory readData,
+            bytes memory targetData
+        ) = _parseOracle(oracleConfig);
```
Comment (Picodes) [inline contracts/transmuter/libraries/LibOracle.sol:29]: _parseOracle -> _parseOracleConfig I'd say

File: README.md [hunk 3]
```diff
@@ -218,13 +218,18 @@ yarn gas
...
-a dummy implementation which has all facets entrypoints. Then go to Etherscan and point the `DiamondProxy` to the `DiamondEtherscanFacet` contract.
-If you happen to update any facet, you need to call `DiamondEtherscanFacet.setDummyImplementation(XXX)` with the newly deployed dummy implementation
+To facilitate the interactions with the Diamond Proxy contract on Etherscan, one solution introduced [here](https://github.com/zdenham/diamond-etherscan/blob/main/README.md) is to deploy the system with a dummy facet that is in fact a noop mock of the whole diamond.
+
+To get the dummy implementation, solution is to:
+
+- download the [repo](https://github.com/zdenham/diamond-etherscan/blob/main/README.md) and follow the instructions
+- upload the dummy implementation [here](./scripts/generated/DummyDiamondImplementation.sol)
+- run [this script](./scripts/gnosis/VerifyProxyEtherscan.s.sol) to deploy the new dummy facet. If the address you're doing this with has ownership on the Transmuter contracts, it will also add the facet to the whole Transmuter system
```
Comment (GuillaumeNervoXS) [inline README.md:229]: You can only do it with gov otherwise it won't do anything

File: contracts/transmuter/facets/AccessControlModifiers.sol [hunk 4]
```diff
@@ -22,24 +22,21 @@ contract AccessControlModifiers {
...
-    // `private` function that does the actual work.
+    /// @notice Prevents a contract from calling itself, directly or indirectly
+    /// @dev This implementation is an adaptation of the OpenZepellin `ReentrancyGuard` for the purpose of this
+    /// Diamond Proxy system. The base implementation can be found here
+    /// https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/security/ReentrancyGuard.sol
     modifier nonReentrant() {
         TransmuterStorage storage ts = s.transmuterStorage();
         // Reentrant protection
-        // On the first call, _notEntered will be true
+        // On the first call, `_notEntered` will be true
```
Comment (GuillaumeNervoXS) [inline contracts/transmuter/facets/AccessControlModifiers.sol:32]: forgot to change - it should be `ts.statusReentrant`

File: scripts/gnosis/VerifyProxyEtherscan.s.sol [hunk 5]
```diff
@@ -49,9 +49,9 @@ contract VerifyProxyEtherscan is Utils {
     // @dev Deploys diamond and connects facets
     function _deployDiamondEtherscan() internal {
         // Deploy every facet
-        facetNames.push("DiamondEtherscanFacet");
-        // etherscanFacet = new DiamondEtherscanFacet();
-        etherscanFacet = DiamondEtherscanFacet(0xC492fBAe68cE6C5E14C7ed5cd8a59babD5c90e4C);
+        facetNames.push("DiamondEtherscan");
+        // etherscanFacet = new DiamondEtherscan();
```
Comment (GuillaumeNervoXS) [inline scripts/gnosis/VerifyProxyEtherscan.s.sol:53]: if you switch comments above we should also switch this one

File: contracts/transmuter/libraries/LibOracle.sol [hunk 6]
```diff
@@ -126,7 +134,15 @@ library LibOracle {
                 );
             }
             return quotePrice;
-        } else return _targetPrice;
+        } else if (readType == OracleReadType.STABLE) return BASE_18;
+        else if (readType == OracleReadType.NO_ORACLE) return baseValue;
+        else if (readType == OracleReadType.WSTETH) return STETH.getPooledEthByShares(1 ether);
```
Comment (GuillaumeNervoXS) [inline contracts/transmuter/libraries/LibOracle.sol:139]: I am not sure I understand why we merge target and read functions? Like all the ETH functions will be useless

Comment (sogipec) [reply contracts/transmuter/libraries/LibOracle.sol:139]: yes they'll be useless for targets, but goal is to only keep one type overall -> so we don't have to do the distinction between normal oracles and targets

File: contracts/transmuter/facets/AccessControlModifiers.sol (hunk 4 above)
Comment (sogipec) [reply contracts/transmuter/facets/AccessControlModifiers.sol]: fixed

File: README.md (hunk 3 above)
Comment (sogipec) [reply README.md]: fixed comment

File: scripts/gnosis/VerifyProxyEtherscan.s.sol (hunk 5 above)
Comment (sogipec) [reply scripts/gnosis/VerifyProxyEtherscan.s.sol]: switched

**Review by GuillaumeNervoXS — APPROVED** [review]

File: test/utils/Transmuter.sol [hunk 7]
```diff
@@ -53,11 +53,6 @@ abstract contract Transmuter is Helper {
         facetNames.push("Swapper");
         facetAddressList.push(address(new Swapper()));
```
Comment (GuillaumeNervoXS) [inline test/utils/Transmuter.sol:55]: why did you remove it?

Comment (sogipec) [reply test/utils/Transmuter.sol:55]: because it's maybe what makes tests so long -> since we have to re-add the facets all the time

File: scripts/DeployTransmuter.s.sol [hunk 8]
```diff
@@ -64,6 +66,10 @@ contract DeployTransmuter is Utils {
         facetNames.push("DiamondCut");
         facetAddressList.push(address(new DiamondCut()));
 
+        // TODO don't forget to set the dummy implementation as an implementation (after having regenerated it)
+        facetNames.push("DiamondEtherscan");
+        facetAddressList.push(address(new DiamondEtherscan()));
+
```
Comment (GuillaumeNervoXS) [inline scripts/DeployTransmuter.s.sol:72]: Can you put back the Redeemer deployment at the very end as this one has trouble being verified on etherscan and the last time it stopped verifiying all the others

File: contracts/transmuter/libraries/LibDiamondEtherscan.sol [hunk 9]
```diff
@@ -2,27 +2,22 @@
...
-     * @dev Storage slot with the address of the current dummy-implementation.
-     * This is the keccak-256 hash of "eip1967.proxy.implementation" subtracted by 1
-     */
-    bytes32 internal constant IMPLEMENTATION_SLOT = 0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc;
-
+    /// @notice Internal version of `setDummyImplementation`
     function setDummyImplementation(address implementationAddress) internal {
-        StorageSlot.getAddressSlot(IMPLEMENTATION_SLOT).value = implementationAddress;
-
+        s.implementationStorage().implementation = implementationAddress;
```
[the remark this answers is not in the document]
Comment (sogipec) [reply contracts/transmuter/libraries/LibDiamondEtherscan.sol:15]: .implementation -> this is how I called the value in the struct
